sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -1339,7 +1340,14 @@ bool coap_block_finish(coap_block_slicer_t *slicer, uint16_t option)
     uint32_t blkopt = _slicer2blkopt(slicer, more);
     size_t olen = _encode_uint(&blkopt);
 
-    coap_put_option(slicer->opt, option - delta, option, (uint8_t *)&blkopt, olen);
+    uint8_t obytes = coap_put_option(slicer->opt, option - delta, option,
+                                     (uint8_t *)&blkopt, olen);
+    if (obytes != slicer->obytes) {
+        /* if the option grew larger, we already corrupted content */
+        assert(obytes < slicer->obytes);
+        memmove(slicer->opt + obytes, slicer->opt + slicer->obytes,
```

Comment:
The length of the move is not correct. Look how the payload changed.
![Screenshot from 2024-05-23 11-31-30](https://github.com/RIOT-OS/RIOT/assets/15147337/7843d16f-caec-408a-b3e5-386c39b11260)

Reply:
arg, that's because the size generated by `coap_opt_add_block2()` is used to calculate the header and therefore the packet length.
Even when updating the `coap_pkt_t` struct in `coap_block2_finish()`, the header length will already have been calculated by `coap_opt_finish()`.